### PR TITLE
Require explict allowlisting of attributes and associations

### LINKED ISSUE #1273: About default searchable attributes and associations
Currently, if you don't restrict the searchable attributes and associations, you can search all ranges.
https://github.com/activerecord-hackery/ransack#authorization-whitelistingblacklisting

I feel that this is very dangerous.
My opinion is exactly the same as the following blog post.
https://younes.codes/posts/how-to-hack-with-ransack

I would like to make this gem secure by default.
Can you please tell me if there is a reason for the current way it is built? Also, I am willing to help if modifications are needed.

Many thanks.

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Fixes #1273.

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
@scarroll32 Ok with releasing this as Ransack 4.0?

File: lib/ransack/adapters/active_record/base.rb [hunk 1]
```diff
@@ -75,6 +70,82 @@ def ransackable_scopes_skip_sanitize_args
...
+              example, `encrypted_password`, `password_reset_token`, `owner` or
+              other sensitive information). You can use the following as a base:
+
+              ```ruby
+              class #{name} < ApplicationRecord
+
+                # ...
+
+                def self.#{method}(auth_object = nil)
+                  #{public_send("unauthorized_#{method}").inspect}
```
Comment (Narnach) [inline lib/ransack/adapters/active_record/base.rb:125]: I think that sorting the names would make it easier to read and thus audit.

```suggestion
                  #{public_send("unauthorized_#{method}").sort.inspect}
```

File: spec/ransack/adapters/active_record/base_spec.rb [hunk 2]
```diff
@@ -689,6 +720,37 @@ def self.simple_escaping?
           it { should include 'parent' }
           it { should include 'children' }
           it { should include 'articles' }
+
+          context 'when not defined in model, nor in ApplicationRecord' do
+            subject { Article.ransackable_associations }
+
+            it "raises a helful error" do
```
Comment (lukas-eu) [inline spec/ransack/adapters/active_record/base_spec.rb:727]: "raises a hel**p**ful error"

File: spec/ransack/adapters/active_record/base_spec.rb [hunk 3]
```diff
@@ -657,6 +657,37 @@ def self.simple_escaping?
             it { should_not include 'only_sort' }
             it { should include 'only_admin' }
           end
+
+          context 'when not defined in model, nor in ApplicationRecord' do
+            subject { Article.ransackable_attributes }
+
+            it "raises a helful error" do
```
Comment (lukas-eu) [inline spec/ransack/adapters/active_record/base_spec.rb:664]: "raises a hel**p**ful error"

File: spec/ransack/adapters/active_record/base_spec.rb (hunk 2 above)
Comment (lukas-eu) [inline spec/ransack/adapters/active_record/base_spec.rb:727]: "raises a hel**p**ful error"

File: lib/ransack/adapters/active_record/base.rb [hunk 4]
```diff
@@ -75,6 +70,82 @@ def ransackable_scopes_skip_sanitize_args
           []
         end
 
+        # Bare list of all potentially searchable attributes. Searchable attributes
+        # need to be explicitly allowlisted through the `ransackable_attributes`
+        # method in each model, but if you're allowing almost everything to be
+        # searched, this list can be used as a base for exclusions.
+        #
+        def unauthorized_ransackable_attributes
```
Comment (lukas-eu) [inline lib/ransack/adapters/active_record/base.rb:78]: `authorizable_ransackable_attributes`?

File: lib/ransack/adapters/active_record/base.rb [hunk 5]
```diff
@@ -75,6 +70,82 @@ def ransackable_scopes_skip_sanitize_args
...
+          end.uniq
+        end
+
+        # Bare list of all potentially searchable associations. Searchable
+        # associations need to be explicitly allowlisted through the
+        # `ransackable_associations` method in each model, but if you're
+        # allowing almost everything to be searched, this list can be used as a
+        # base for exclusions.
+        #
+        def unauthorized_ransackable_associations
```
Comment (lukas-eu) [inline lib/ransack/adapters/active_record/base.rb:93]: `authorizable_ransackable_associations`?

**Review by lukas-eu — COMMENTED** [review]
A typo and another idea for naming the list of all attributes/associations

File: lib/ransack/adapters/active_record/base.rb (hunk 4 above)
Comment (deivid-rodriguez) [reply lib/ransack/adapters/active_record/base.rb:78]: So much better, thank you!

File: lib/ransack/adapters/active_record/base.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/ransack/adapters/active_record/base.rb:125]: Makes sense, thanks!

**Comment by deivid-rodriguez** [comment]
I will prepare a release tomorrow.

**Comment by deivid-rodriguez** [comment]
@scarroll32 I created a release draft. It's been a while without releases, so let me know how it looks if you have some time! Otherwise I'll just release this tommorrow™️.

**Comment by scarroll32** [comment]
References

- https://positive.security/blog/ransack-data-exfiltration
- https://portswigger.net/daily-swig/ruby-on-rails-apps-vulnerable-to-data-theft-through-ransack-search

**Comment by AndreyCerqueiraLima** [comment]
I really understood the security idea about this change.

But when something doesn't have a door the path is "free". 
When you put a door on the path and force every player to implement a key to use this door, everything is okay when this door is just for 2 people. 

However, when all of the world need to use this door, you generate a big problem! Imagine, need to generate a key for everyone.
 
So, this behavior break so many applications that discourages people use ransack. Because the core changed and break the whole app, forced to implement this security feature.

**Comment by night91** [comment]
Migrating a full app tying to get all attributes and asossiations you use a a big deal. Is there a workaround? I have not been able to migrate to 4 and I cant not migrate to rails 7.2 and 8.0 from 3.0.

**Pull request merged**